doc/sphinx/proof-engine/proof-handling.rst

```diff
@@ -539,6 +539,112 @@ Requesting information
    fixpoint and cofixpoint is violated at some time of the construction
    of the proof without having to wait the completion of the proof.
 
+.. _showing_diffs:
+
+Showing differences between proof steps
+---------------------------------------
+
+
+Coq can automatically highlight the differences between successive proof steps.
+For example, the following screen shots of coqtop and CoqIDE show the same
+“intros” tactic.  This tactic creates two new hypotheses, highlighted in green.
+The conclusion is entirely in pale green because it’s changed, but no tokens were added
+to it.  The third screen shot uses the "removed" option, so it shows the conclusion a
+second time with the old text, with deletions marked in red.  Also, since the hypotheses are
+new, no line of old text is shown for them.
+
+.. _diffs_coqtop_on:
+
+  .. image:: ../_static/diffs-coqtop-on.png
+     :alt: coqtop with Set Diffs on
+
+.. _diffs_coqide_on:
+
+  .. image:: ../_static/diffs-coqide-on.png
+     :alt: |CoqIDE| with Set Diffs on
+
+.. _diffs_coqide_removed:
+
+  .. image:: ../_static/diffs-coqide-removed.png
+     :alt: |CoqIDE| with Set Diffs removed
+
+**How to enable diffs**
+
+.. opt:: Set Diffs %( "on" %| "off" %| "removed" %)
+.. This ref doesn't work: :opt:`Set Diffs %( "on" %| "off" %| "removed" %)`
+
+The “on” option highlights added tokens in green, while the “removed” option
+additionally reprints items with removed tokens in red.  Unchanged tokens in
+modified items are shown with pale green or red.  (Colors are user-configurable.)
+
+For coqtop, showing diffs can be enabled when starting coqtop with the
+``-diffs on|off|removed`` command-line option or with the ``Set Diffs``
+command within Coq.  You will need to provide the ``-color on|auto`` command-line option when
+you start coqtop in either case.
+
+Colors for coqtop can be configured by setting the ``COQ_COLORS`` environment
+variable, which has the same format as the GNU ``LS_COLORS`` variable.  The string
+uses ANSI escape codes to represent attributes.  For example:
+
+        ``export COQ_COLORS=”diff.added=4;48;2;0;0;240:diff.added.bg=44”``
+
+sets the highlights for added text to underlined (the 4) with a background RGB
+color (0, 0, 240) and unmodified text in a modified item to a blue background.
+The other tags used by diffs are ``diff.removed`` and ``diff.removed.bg``.  Note
+that if you specify ``COQ_COLORS``, the predefined attributes are ignored.
+
+
+In CoqIDE, diffs should be enabled from the View menu.  Don’t use the ``Set Diffs``
+command in CoqIDE.  You can change the background colors shown for diffs from the
+Edit | Preferences | Tags panel by changing the settings for the ``diff.added``,
+``diff.added.bg``, ``diff.removed`` and ``diff.removed.bg`` tags.  This panel also
+allows let you control other attributes of the highlights, such as the foreground
+color, bold, italic, underline and strikeout.
+
+Note: As of this writing (August 2018), Proof General will need minor changes
```

Comment:
Use `.. note::`
BTW, is there an open issue on PG issue tracker? You could point to it.

Reply:
Right, it would be nice to know which minor adjustments are needed.  Even better, it would be nice to implement the feature in PG before merging this, and then we don't even need to document the issue.

Reply:
I'll open an issue in PG and reference it.  I was told there are 2 versions: a general PG and a Coq-specific PG.  Is https://github.com/ProofGeneral/PG the correct project?

I assumed I'd get better attention from the PG folks after the feature is officially released.  I've not interacted with the PG folks (except a little with Erik Dorel-Martin, who I met in Nice) so I don't know if or when they would make changes.  Suggestions welcome.

> Right, it would be nice to know which minor adjustments are needed.

Presumably in the PR, not here.

> it would be nice to implement the feature in PG before merging this

You make it sound so easy.  I've never even used PG, I'm not an emacs user, haven't done emacs lisp and I don't know how the PG team operates.  Maybe none of these will present any difficulties but I don't know yet.  Still busy making sure the last few changes get into 8.9.

Reply:
There is only one version and it is Coq-specific and it is the right URL you liked to.

>I assumed I'd get better attention from the PG folks after the feature is officially released.

That's a bad idea: we generally try to have PG support land as early as possible. Note that @cpitclaudel knows what he's talking about since he's also a PG maintainer.

>You make it sound so easy. I've never even used PG, I'm not an emacs user, haven't done emacs lisp and I don't know how the PG team operates. Maybe none of these will present any difficulties but I don't know yet. Still busy making sure the last few changes get into 8.9.

But do you know for sure that minor adjustments are needed? I didn't try it yet.

Reply:
Well, I spoke to Erik about supporting diffs in PG when I was in Nice, then I emailed him once the diff code was in master.  I didn't hear back from him.  I would be delighted if one of the PG maintainers would make the changes.

Adjustments needed (I'll put more detail into the PR):
1. Define default color/highlight for tags diff.removed, diff.added, diff.removed.bg, diff.added.bg
2. Semantic extension: tags in the form "start.*" indicate the highlight continues until a matching "end.*" tag is seen.  But turn off such highlighting for white space at the beginning and end of lines.

Reply:
PS PG will work just fine if diffs are not enabled.  It's only if diffs are enabled.

Reply:
Looks pretty good to me with -emacs mode.  @ejgallego advised me to change Idetop.goals, which IIUC is for the XML interface, which I also did.  If PG simply accepts ANSI escape codes embedded in the output, then it's probably really simple.  Can PG/emacs handle underlining (or strikeout?).  I would expect it can handle background colors with 24-bit rgb colors.  If it can handle strikeout, I may want to make that available in coqtop--I had read that most terminal software doesn't support strikeout.

Also, the colors I chose may not be great defaults in emacs.  Is there a good way to provide different defaults from PG?  Or is color selection usually done just by setting COQ_COLORS?

Is there already clean way to pass `-diffs "on"` from PG so the user doesn't have to do that every time they start PG (if they want that setting)?  Or is that a small change?

Perhaps you can try the code in master with PG with -emacs and let me know how it looks?

![image](https://user-images.githubusercontent.com/1253341/44547670-bc751c00-a6d0-11e8-90d1-69d6df10beb1.png)

Reply:
>  If PG simply accepts ANSI escape codes embedded in the output, then it's probably really simple.

We don't currently do, AFAIK, but it could be done.

> Can PG/emacs handle underlining (or strikeout?).

I expect so.

>  I would expect it can handle background colors with 24-bit rgb colors.

Not on typical terminals (these tend to support only 256 colors), though the latest versions of Emacs can be configured to output 24-bit color codes on appropriately configured terminals.  
Of course, the GUI Emacs doesn't have trouble with 24-bit color.

> Also, the colors I chose may not be great defaults in emacs. Is there a good way to provide different defaults from PG? Or is color selection usually done just by setting COQ_COLORS?

We currently ignore the colors produced by coqtop (or, more precisely, colors are currently disabled with -emacs).  You mentioned 24-bit colors; do you actually make use of those, or do you use the more typical 256-color terminal palette?

Can you show what the raw output (non-highlighted) looks like?

Reply:
The coqtop colors are 24-bit.  It was tricky to get shades of green and red that worked reasonably well with a black background and without changing the existing syntactic highlights.

```
  ; "diff.added"       , make ~bg_color:(`RGB(0,141,0)) ~underline:true ()
  ; "diff.removed"     , make ~bg_color:(`RGB(170,0,0)) ~underline:true ()
  ; "diff.added.bg"    , make ~bg_color:(`RGB(0,91,0)) ()
  ; "diff.removed.bg"  , make ~bg_color:(`RGB(91,0,0)) ()
```

The CoqIDE colors are exactly the ones that GitHub uses, which I think are better than the coqtop colors.  Depends on whether PG usually has a light or dark background.  They require 24 bit color:

```
    ("diff.added", make_tag ~bg:"#b6f1c0" ~underline:true ());
    ("diff.removed", make_tag ~bg:"#f6b9c1" ~strikethrough:true ());
    ("diff.added.bg", make_tag ~bg:"#e9feee" ());
    ("diff.removed.bg", make_tag ~bg:"#fce9eb" ());
```

Here is the raw output, omitting the escape character that should appear before each `[`:

```$ rlwrap bin/coqtop.exe -diffs removed -emacs -color on
Welcome to Coq DESKTOP-MMPNVGH:/cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq,diffs_doc (c66c27781c555ec7301300cbf0d0342394c03981)

<prompt>Coq < 1 || 0 < </prompt>Parameter P : nat -> Prop.
<infomsg>P is declared</infomsg>

<prompt>Coq < 2 || 0 < </prompt>Goal P 1 /\ P 2.
1 subgoal (ID 1)

  ============================
  [48;2;0;91;0m[48;2;0;141;0;4mP 1[37m /\[39m P 2[48;2;0;91;0;24m

<prompt>Unnamed_thm < 3 |branch| 0 < </prompt>split.
2 subgoals (ID 3)

  ============================
  [48;2;91;0;0mP 1[37m [39m[48;2;170;0;0;4m[37m/\[39m P 2[48;2;91;0;0;24m
  [48;2;0;91;0mP 1

subgoal 2 (ID 4) is:
 P 2
```

Reply:
Thanks, I looked into this.  The parser in Emacs doesn't support these 24-bit color codes :/
Adding them shouldn't be too hard, but I don't really have the bandwidth for that at the moment.
Is there a way to specify COQ_COLORS that would make Coq use the more common 8-color palette? It includes standard and bright red and green, so we could use this for Emacs.

Reply:
Yes, setting this in COQ_COLORS should work just fine.  Tag names are shown above.

Should I still submit an issue for PG?  Seems like you may be able to explain what's needed at this point better than I can.  If you create the issue, please make sure I see it so I can comment on it.

Reply:
Done at https://github.com/ProofGeneral/PG/issues/381 . Thanks for taking the time to explain!

Reply:
Thanks for your help!